keps/sig-node/4639-oci-volume-source/README.md

```diff
@@ -0,0 +1,1009 @@
+   - Ensure that pulling and storing these artifacts is as efficient and secure as with OCI images.
+
+**Lifecycling and Garbage Collection:**
+   - Reuse the existing kubelet logic for managing the lifecycle of OCI objects.
+
+**Artifact-Specific Configuration:**
+   - Introduce new configuration options to handle the unique requirements of different types of OCI artifacts.
+
+**Artifacts as Subject Referrers:**
+   - Introduce new refer to image and filter for artifact type criterion/options for to be mounted artifact(s).
+   - Certain types of OCI artifacts include a subject reference. That reference
+     identifies the artifact/image for which this artifact refers. For example a
+     signature artifact could refer to a platform index, for certifying the
+     platform images, or to an SBOM artifact that refers to a platform matched
+     image. These artifacts may or may not be located on the same
+     registry/repository. The new referrers API allows for retrieving these
+     artifacts.
+
+**Validation:**
+   - Extend validation and security checks to cover new artifact types.
+
+**Storage Optimization in the container runtime:**
+   - Develop optimized storage solutions tailored for different artifact types,
+     potentially integrating with existing storage solutions or introducing new mechanisms.
+
+#### kubelet
+
+While the container runtime will be responsible of pulling and storing the OCI
+objects in the same way as for images, the kubelet still has to manage the full
+life-cycle of them. This means that some parts of the existing kubelet code can
+be reused, for example:
+
+- The logic how to ensure that an image exists on the node:
+  https://github.com/kubernetes/kubernetes/blob/39c6bc3/pkg/kubelet/images/image_manager.go#L102
+- The retrieval of available secrets for a pod:
+  https://github.com/kubernetes/kubernetes/blob/39c6bc3/pkg/kubelet/kubelet_pods.go#L988
+
+Specifying a `pullPolicy` be supported in the same way as for the existing
+`imagePullPolicy` API.
+
+For registry authentication purposes the same logic will be used as for the
+container image.
+
+#### CRI
+
+The CRI API is already capable of managing container images [via the `ImageService`](https://github.com/kubernetes/cri-api/blob/3a66d9d/pkg/apis/runtime/v1/api.proto#L146-L161).
+Those RPCs will be re-used for managing OCI artifacts, while the [`ImageSpec`](https://github.com/kubernetes/cri-api/blob/3a66d9d/pkg/apis/runtime/v1/api.proto#L798-L813)
+as well as [`PullImageResponse`](https://github.com/kubernetes/cri-api/blob/3a66d9d/pkg/apis/runtime/v1/api.proto#L1530-L1534)
+will be extended to mount the OCI object to a local path:
+
+```protobuf
+
+// ImageSpec is an internal representation of an image.
+message ImageSpec {
+    // …
+
+    // Indicate that the OCI object should be mounted.
+    bool mount = 20;
+
+    // SELinux label to be used.
+    string mount_label = 21;
+}
```

Comment:
Are there any `fsGroup` or `uid` based permissions to consider?

Reply:
I don't think so, that's handled by the runtime providing the mount. What would be the use case?

Reply:
Is there a upper limit to image sizes? My concern is - if there are large images with lots of files, runtime is recursively both changing selinux perms and file based permissions and then we will hear complaints from customers that pods that are using these images are failing to start. 

We have worked around this via APIs in kubernetes to some extent. (See selinux and fsgroup related KEPs). Are these solved problems for CRI? Will these volume types use any of existing APIs that control fsgroup and selinux handling in pods from storage POV? cc @jsafrane

Reply:
What use cases would there be for a custom label?  I think the mount should always use the same label type as that for the base pod image.

> Is there a upper limit to image sizes? My concern is - if there are large images with lots of files, runtime is recursively both changing selinux perms and file based permissions and then we will hear complaints from customers that pods that are using these images are failing to start.

Note that these are not *mutable* volumes, so there's no need to chown anything ever. I think in practice the default would be that the files in the volume are root owned and world readable, that's it.

Reply:
@cgwalters yes, my idea was to use the `mount_label` field in the CRI inherited from the pod. Then use `mount -o context=…` in the runtime (there is an API) to read-only mount with the target label.

Reply:
Will this `VolumeSource` support `subPaths` ? If yes - using `-o context` may not always work....

Reply:
It exists as an API in pod's `volumeMount` field. I am not sure if there is a way to block it. What will happen if different containers in the pod are trying to mount different `subPath`'s from the image?

Reply:
> It exists as an API in pod's volumeMount field. I am not sure if there is a way to block it.

The API validation should be able to do that: https://github.com/kubernetes/kubernetes/blob/efef32652af0af08a0b9c9bc547a4dce4a95f9f5/pkg/apis/core/validation/validation.go#L3613

> What will happen if different containers in the pod are trying to mount different subPath's from the image?

That will be forbidden.